### PR TITLE
GitHub Issue NOAA-EMC/GSI#375.  Improvements to minimization

### LINKED ISSUE #375: Improvements to minimization
Changes are in branch converge.

### PR DESCRIPTION
**Pull request by jderber-NOAA**
Ref: Issue # 375

This change modifies GSI code to use sensible temperature rather than virtual temperature in the rad routines and fixes a minor bug in the radiance and jacobian calculations in setuprad.  These changes do not change results much.

Also, it introduces a new namelist variable optconv.  The optconv variable when set greater than zero (recommended values for global 0.06) down weights channels based on their sensitivity to moisture to improve the convergence characteristics of the minimization.  See issue #375 for details.

### CODE REVIEW

**Comment by MichaelLueken** [comment]
Hi @jderber-NOAA, there is an issue that needs to be addressed.  There are 32 commits associated with this PR and it needs to be reduced to a single commit.  To do this:

- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git stash
- git remote add upstream https://github.com/NOAA-EMC/GSI.git
- git remote update
- git rebase upstream/develop
- git stash pop
- Please correct the conflicts in src/gsi/crtm_interface.f90 and src/gsi/gsimod.F90 (keep stashed changes)
- git add src/gsi/crtm_interface.f90 src/gsi/gsimod.F90
- `git commit -m "GitHub Issue NOAA-EMC/GSI#375.  Improvements to minimization."`
- git push origin converge --force

This will reduce the number of commits from 32 to 1 and ensure that your PR is up-to-date with the current authoritative develop branch.  Please let me know if you have any questions or encounter any issues.

File: src/gsi/crtm_interface.f90 [hunk 1]
```diff
@@ -1697,21 +1697,16 @@ subroutine call_crtm(obstype,obstime,data_s,nchanl,nreal,ich, &
...
-        qsat(k)=(ges_qsat (ix ,iy ,k, itsig)*w00+ &
-                 ges_qsat (ixp,iy ,k, itsig)*w10+ &
-                 ges_qsat (ix ,iyp,k, itsig)*w01+ &
-                 ges_qsat (ixp,iyp,k, itsig)*w11)*dtsig + &
-                (ges_qsat (ix ,iy ,k, itsigp)*w00+ &
-                 ges_qsat (ixp,iy ,k, itsigp)*w10+ &
-                 ges_qsat (ix ,iyp,k, itsigp)*w01+ &
-                 ges_qsat (ixp,iyp,k, itsigp)*w11)*dtsigp
-     end if
-     c2(k)=one/(one+fv*q(k))
```
Comment (MichaelLueken) [inline src/gsi/crtm_interface.f90:1710]: While compiling in debug mode, fv was identified as being unused.  Please remove fv from line 1049.

File: src/gsi/setuprad.f90 [hunk 2]
```diff
@@ -1771,34 +1774,46 @@ subroutine setuprad(obsLL,odiagLL,lunin,mype,aivals,stats,nchanl,nreal,nobs,&
...
+             raterr2(ii)=error0(ii)**2*varinv0(ii)
+             if (l_may_be_passive .and. .not. retrieval) then
+              if(optconv > zero .and. (iasi .or. cris) .and. iinstr .ne. -1)then
+               asum=zero
+               do k=1,nsig
+                  asum=asum+abs(jacobian(iqs+k,ii))*qs(k)
+               end do
+               diagadd(ii)=optconv*asum
+              end if
+             end if
```
Comment (MichaelLueken) [inline src/gsi/setuprad.f90]: Please ensure at least two-space indentation for if statements and do loops.  Please replace .ne. with /= on line 1788.

**Comment by MichaelLueken** [comment]
@ADCollard @emilyhcliu

I know that you are both focused on the GFS v16.3.0 implementation right now, so I wanted to reach out and see if I need to work with John to find someone else to review these changes.  Please let me know as soon as possible.  Thanks!

File: src/gsi/crtm_interface.f90 [hunk 3]
```diff
@@ -1806,7 +1801,7 @@ subroutine call_crtm(obstype,obstime,data_s,nchanl,nreal,ich, &
 ! if ( icmask .and. n_clouds_fwd_wk > 0 .and. imp_physics==11 .and.  lcalc_gfdl_cfrac ) then
   if ( icmask .and. n_clouds_fwd_wk > 0 .and. imp_physics==11 .and.  lprecip_wk ) then
      cf_calc  = zero
-     call calc_gfdl_cloudfrac(rho_air,h,q,cloud,hs,garea,qsat,cf_calc)
```
Comment (MichaelLueken) [inline src/gsi/crtm_interface.f90:1809]: While compiling in debug mode, qsat was identified as unused.  Please remove qsat from line 1115.

**Review by ADCollard — APPROVED** [review]
I am sorry this review was delayed.   I think these changes look good and the solution to the minimisation by inflating the the errors based on the size of the humidity Jacobian is a nice idea.  I will test this further in a parallel experiment as I think this could potentially solve some issues we are currently having with the CrIS performance.

**Review by MichaelLueken — CHANGES_REQUESTED** [review]
Hi @jderber-NOAA.  I have completed my visual inspection of the changes.  For the most part, everything looks good.  Please ensure that lines 1788 through 1794 are indented at least two spaces in src/gsi/setuprad.f90.  Also, please replace .ne. with /= on line 1788 in src/gsi/setuprad.f90, as well.  Additionally, while compiling in debug mode, two unused variables were identified.  Please remove unused variables fv and qsat, from lines 1049 and 1115, respectively, in src/gsi/crtm_interface.f90.

Once these changes have been made, please use:

- git add crtm_interface.f90 setuprad.f90
- git commit --amend
- Please save and close the commit message window without changing the commit message.
- git push origin converge --force

Please let me know if you have any questions or encounter any issues.

**Review by MichaelLueken — APPROVED** [review]
Hi @jderber-NOAA.  Thank you very much for making these changes!  I will submit this work to the review committee once the work that is currently out has been merged.

**Comment by MichaelLueken** [comment]
The due date for feedback from the review committee has past, so I will now give final approval to this update and merge it to the authoritative develop branch.

**Pull request merged**